### PR TITLE
JAMES-2586 Fix postgres mailbox acl concurrency issue

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

**Review by Arsnael — COMMENTED** [review]
Read it, same remark

**Review by quantranhong1999 — COMMENTED** [review]
Read it.

**Comment by hungphan227** [comment]
> No test?

added

**Review by Arsnael — APPROVED** [review]

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMailboxMapper.java [hunk 1]
```diff
@@ -96,21 +95,21 @@ public Flux<Mailbox> findNonPersonalMailboxes(Username userName, MailboxACL.Righ
...
     public Mono<ACLDiff> updateACL(Mailbox mailbox, MailboxACL.ACLCommand mailboxACLCommand) {
-        return upsertACL(mailbox,
-            mailbox.getACL(),
-            Throwing.supplier(() -> mailbox.getACL().apply(mailboxACLCommand)).get());
+        return postgresMailboxDAO.getACL(mailbox.getMailboxId())
+            .flatMap(pairMailboxACLAndVersion -> {
+                try {
+                    MailboxACL newACL = pairMailboxACLAndVersion.getLeft().apply(mailboxACLCommand);
+                    return postgresMailboxDAO.upsertACL(mailbox.getMailboxId(), newACL, pairMailboxACLAndVersion.getRight())
+                        .thenReturn(ACLDiff.computeDiff(pairMailboxACLAndVersion.getLeft(), newACL));
```
Comment (vttranlina) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMailboxMapper.java:103]: The generate `ACLDiff` base on result of query `upsertACL` will more reliable

My mean:
```
upsertACL(xxx)
.map(newAclResult -> calculateACLDiff(newAclResult)))
```

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxDAO.java [hunk 2]
```diff
@@ -150,15 +151,34 @@ private Mono<MailboxId> update(Mailbox mailbox) {
...
             .then();
     }
 
+    public Mono<Void> upsertACL(MailboxId mailboxId, MailboxACL acl) {
+        return postgresExecutor.executeReturnAffectedRowsCount(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+                .set(MAILBOX_ACL, MAILBOX_ACL_TO_HSTORE_FUNCTION.apply(acl))
+                .set(DSL.field(MAILBOX_ACL_VERSION.getName()), (Object) DSL.field(MAILBOX_ACL_VERSION.getName() + " + 1"))
+                .where(MAILBOX_ID.eq(((PostgresMailboxId) mailboxId).asUuid()))))
+            .filter(count -> count > 0)
+            .switchIfEmpty(Mono.error(new RuntimeException("Upsert mailbox acl failed with mailboxId " + mailboxId.serialize())))
```
Comment (quantranhong1999) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxDAO.java:171]: IMO we should add retries. And BTW the error seems not clear about the root cause to me.

Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxDAO.java:172]: added

**Review by chibenwa — APPROVED** [review]
Ok to me

**Comment by Arsnael** [comment]
https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2551/3/testReport/junit/org.apache.james.jmap.rfc8621.postgres/PostgresMailboxSetMethodTest/lookupRightShouldBePropagated_GuiceJamesServer_/

Related?

**Review by Arsnael — APPROVED** [review]

**Pull request merged**